staging/src/k8s.io/apiserver/pkg/admission/attributes.go

```diff
@@ -140,6 +145,70 @@ func (record *attributesRecord) AddAnnotation(key, value string) error {
 	return nil
 }
 
+func (record *attributesRecord) GetReinvocationContext() ReinvocationContext {
+	return record.reinvocationContext
+}
+
+func newReinvocationContext() *reinvocationContext {
+	return &reinvocationContext{previouslyInvokedRevocableWebhooks: sets.NewString(), reinvokeWebhooks: sets.NewString()}
+}
+
+type reinvocationContext struct {
+	// isReinvoke is true when admission plugins are being reinvoked
+	isReinvoke bool
+	// lastOutput holds the result of the last webhook admission plugin call
+	lastOutput runtime.Object
+	// previouslyInvokedRevocableWebhooks holds the set of webhooks that have been invoked and
+	// should be reinvoked if a later mutation occurs
+	previouslyInvokedRevocableWebhooks sets.String
+	// reinvokeWebhooks holds the set of webhooks that should be reinvoked
+	reinvokeWebhooks sets.String
+	// reinvokeInTree indicates in-tree plugins should be reinvoked
+	reinvokeInTree bool
+}
+
+func (rc *reinvocationContext) IsReinvoke() bool {
+	return rc.isReinvoke
+}
+
+func (rc *reinvocationContext) SetIsReinvoke() {
+	rc.isReinvoke = true
+}
+
+func (rc *reinvocationContext) ShouldReinvoke() bool {
+	return rc.reinvokeInTree || len(rc.reinvokeWebhooks) > 0
+}
+
+func (rc *reinvocationContext) IsOutputChangedSinceLastWebhookInvocation(object runtime.Object) bool {
+	return !apiequality.Semantic.DeepEqual(rc.lastOutput, object)
+}
+
+func (rc *reinvocationContext) SetLastWebhookInvocationOutput(object runtime.Object) {
+	if object == nil {
+		rc.lastOutput = nil
+		return
+	}
+	rc.lastOutput = object.DeepCopyObject()
+}
+
+func (rc *reinvocationContext) ShouldInvokeWebhook(webhook string) bool {
+	return !rc.isReinvoke || rc.reinvokeWebhooks.Has(webhook)
+}
+
+func (rc *reinvocationContext) AddReinvocableWebhookToPreviouslyInvoked(webhook string) {
+	rc.previouslyInvokedRevocableWebhooks.Insert(webhook)
+}
+
+func (rc *reinvocationContext) RequireReinvokingPreviouslyInvokedPlugins() {
+	if len(rc.previouslyInvokedRevocableWebhooks) > 0 {
+		for s := range rc.previouslyInvokedRevocableWebhooks {
+			rc.reinvokeWebhooks.Insert(s)
```

Comment:
rc.reinvokeWebhooks.Insert(rc.previouslyInvokedRevocableWebhooks...)

Reply:
Since rc.previouslyInvokedRevocableWebhooks is a `sets.String` (`map[string]Empty`) it doesn't expand via `...`.

Reply:
List() gives you the list.

I would have bet that we have some set variant which is just a sorted slice.

Reply:
I suppose if our goal is readability `rc.reinvokeWebhooks = rc.reinvokeWebhooks.Union(rc.previouslyInvokedRevocableWebhooks)` might be the best? I had gone with `Insert()` because it's the cheapest way to do the operation, but I don't feel strongly about it since the expected size of the sets are small.